[NFC][flang][OpenMP] Create `FortranUtils` lib and move `createMapInfoOp` to it

ergawy opened the pull request:
This moves `createMapInfoOp` from `flang/lib/Lower/OpenMP/Utils.h` to new utils library.

Context: I am working on upstreaming (from AMD's downstream fork) support for `do concurrent` mapping to the GPU. This means that `DoConcurrentConversion.cpp` needs access to `createMapInfoOp`. Hence, we moved it downstream to a shared location between that is linked by both `FlangOpenMPTransforms` (where the `do concurrent` pass lives) and `FortranLower` (where `createMapInfoOp` originally were).

Note that, so far, upstream `DoConcurrentConversion.cpp` does not reference `createMapInfoOp` yet. Follow-up PRs will upstream this later.

Commit:
[flang][OpenMP] Move `createMapInfoOp` to `OpenMP-utils.h`

This moves `createMapInfoOp` from `flang/lib/Lower/OpenMP/Utils.h` to `flang/include/flang/Support/OpenMP-utils.h`.

Context: I am working on upstreaming (from AMD's downstream fork) support for `do concurrent` mapping to the GPU. This means that `DoConcurrentConversion.cpp` needs access to `createMapInfoOp`. Hence, we moved it downstream to a shared location between that is linked by both `FlangOpenMPTransforms` (where the `do concurrent` pass lives) and `FortranLower` (where `createMapInfoOp` originally were).

The issue now is that we have to link in both the FIR and OpenMP MLIR dialects which is not ideal to link with a suport library like `FortranSupport`.

Note that, so far, upstream `DoConcurrentConversion.cpp` does not reference `createMapInfoOp` yet. Follow-up PRs will upstream this later.

Meinersbur commented on flang/include/flang/Support/OpenMP-utils.h:
```diff
@@ -72,6 +73,14 @@ struct EntryBlockArgs {
 mlir::Block *genEntryBlock(
     mlir::OpBuilder &builder, const EntryBlockArgs &args, mlir::Region &region);
+
+mlir::omp::MapInfoOp createMapInfoOp(mlir::OpBuilder &builder,
```
Consider adding some `///` documentation

Meinersbur commented:
Since the PR ios for discussion, I am not LGTM it yet. Consider adding NFC to the title.

Commit:
Add docs

ergawy commented:
@Meinersbur Thanks for the review. Added docs. Also added the `[NFC]` tag, I hope that does not jinx the PR :D.

ergawy commented:
Ping! Please take a look and let me know if you have objections to changes in this PR.

tblah reviewed (commented):
This is okay with me so long as you're sure the new dependencies in fir/lib/Support/CMakeLists.txt do not introduce circular dependencies. I presume any circular dependencies would have already broken builds for your downstream?

So I guess the important change is adding a dependency on FIRDialect to flang/Support. So far as I can tell this looks safe.

mjklemm reviewed (commented):
Moving this to shared place seems to make sense to avoid code duplication.

bhandarkar-pranav reviewed (commented):
@ergawy  - Is now a good time to move `genImplicitBounds` and its helpers from [here](https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Builder/DirectivesCommon.h#L332) too? Does `DoConcurrent` not use it?

TIFitis reviewed (commented):
Could we have `MapInfoFinalization`, _flang/lib/Optimizer/OpenMP/MapsForPrivatizedSymbols.cpp_, etc make use of this? Having a single PR to refactor all of these seems like a good idea to me.

agozillon reviewed (approved):
LGTM, please do make sure this builds with -DBUILD_SHARED_LIBS=ON and OFF! Just recalling some minor issues from this refactor downstream that'd be good to avoid repeating upstream :-)

bhandarkar-pranav commented:
> > @ergawy - Is now a good time to move `genImplicitBounds` and its helpers from [here](https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Builder/DirectivesCommon.h#L332) too? Does `DoConcurrent` not use it?
> 
> Thought I'd mention I have a PR open for `genImplicitBounds` here - #154164.

Yikes, sorry for missing this @TIFitis

ergawy commented:
> LGTM, please do make sure this builds with -DBUILD_SHARED_LIBS=ON and OFF! Just recalling some minor issues from this refactor downstream that'd be good to avoid repeating upstream :-)

Tried both with an empty build folder and builds and lit tests work fine.

ergawy commented:
> > The `Support` library in llvm is probably a good example to follow in what should we add here or not.
> 
> Can you please elaborate on this? Is there anything that we can imitate here?

@abidh static libs + debug work (builds and lit tests). dynamic libs + debug fails to link even without the changes in this PR (on `main`):
```
/home/kaergawy/git/aomp21.0/llvm-project/flang/include/flang/Common/indirection.h:88: undefined reference to `Fortran::common::die(char const*, ...)'
/usr/bin/ld: tools/flang/lib/Optimizer/Builder/CMakeFiles/FIRBuilder.dir/IntrinsicCall.cpp.o: ....
```

Meinersbur commented:
I think there is some usefulness. "Support" accross LLVM projects denotes general purpose utilities that could have been part of the C++ standard library, i.e. are not even specific to a compiler project[^1]. IMHO `llvm::Triple` never fit into Support, since it is specific to a compiler project.

[^1]: `clangSupport` seems to be the [exception](https://github.com/llvm/llvm-project/commit/f26c41e8dd28d86030cd0f5a6e9c11036acea5d2) and [should not have been called "Support"](https://reviews.llvm.org/D121984#3440277). It's creation has the same background as LLVMTargetParser, bit in the end discussing its name is equivalent to bikeshedding.

Commit:
Add `FortranUtils` libarary.

ergawy commented:
> I think there is some usefulness. "Support" accross LLVM projects denotes general purpose utilities that could have been part of the C++ standard library, i.e. are not even specific to a compiler project[1](#user-content-fn-1-e5797b45201fb769eaa791c81c3bcf1c). IMHO `llvm::Triple` never fit into Support, since it is specific to a compiler project.
> ## Footnotes
> 
>     1. `clangSupport` seems to be the [exception](https://github.com/llvm/llvm-project/commit/f26c41e8dd28d86030cd0f5a6e9c11036acea5d2) and [should not have been called "Support"](https://reviews.llvm.org/D121984#3440277). It's creation has the same background as LLVMTargetParser, bit in the end discussing its name is equivalent to bikeshedding. [↩](#user-content-fnref-1-e5797b45201fb769eaa791c81c3bcf1c)

Thanks for the context. I added a new lib and moved the function to it.

Commit:
add todo

abidh commented:
> > > > The `Support` library in llvm is probably a good example to follow in what should we add here or not.
> > > 
> > > 
> > > Can you please elaborate on this? Is there anything that we can imitate here?
> > 
> > 
> > @abidh static libs + debug work (builds and lit tests). dynamic libs + debug fails to link even without the changes in this PR (on `main`):
> > ```
> > /home/kaergawy/git/aomp21.0/llvm-project/flang/include/flang/Common/indirection.h:88: undefined reference to `Fortran::common::die(char const*, ...)'
> > /usr/bin/ld: tools/flang/lib/Optimizer/Builder/CMakeFiles/FIRBuilder.dir/IntrinsicCall.cpp.o: ....
> > ```
> 
> I will investigate this failure.

It seems that it got fixed in https://github.com/llvm/llvm-project/pull/155422.

ergawy commented:
Thanks for looking into the linking failure. Let me know if we can merge the PR now or still have other comments.

abidh reviewed (approved):
Thanks for working on it. LGTM.
